src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxNode.cs

```diff
@@ -91,21 +91,25 @@ private static void ComputeSyntaxTree(CSharpSyntaxNode node)
                 var parent = node.Parent;
                 if (parent == null)
                 {
-                    Interlocked.CompareExchange(ref node._syntaxTree, CSharpSyntaxTree.CreateWithoutClone(node), null);
-                    tree = node._syntaxTree;
+                    // set the tree on the root node atomically
+                    tree = CSharpSyntaxTree.CreateWithoutClone(node);
+                    var orig = Interlocked.CompareExchange(ref node._syntaxTree, tree, null);
+                    if (orig != null)
+                    {
+                        tree = orig;
+                    }
                     break;
                 }
-                else if (parent._syntaxTree != null)
+
+                tree = parent._syntaxTree;
+                if (tree != null)
                 {
-                    Interlocked.CompareExchange(ref node._syntaxTree, parent._syntaxTree, null);
-                    tree = node._syntaxTree;
+                    node._syntaxTree = tree;
```

Comment:
it would be the same tree. In a worst and unlikely case I assign it more than once.
We are in a huge trouble if we could have different trees here.

Use of CmpExch was expensive in this method according to profiler. I removed it where it was not adding uniqueness guarantees.